### PR TITLE
✨ Add annotation support to util.CloneTemplate to pass them down from templates to machines

### LINKED ISSUE #4543: Pass down annotations from Machine to provider specific machine resources
**User Story**

As a user I'd like to specify annotations as part of a MachineDeployment and use them on the provider specific machine resources to pass parameters to a custom controller that assigns static IP addresses.

**Detailed Description**

It's possible to specify both labels and annotations as part of the `spec.template.metadata` of a `MachineDeployment`. Both get passed down to the `MachineSet` and subsequently the `Machine` resources.
The provider specific machine resource only gets the labels, and not the annotations.

Annotations specified on `Machine`s should be passed down to the provider specific resources, which is currently only done for labels.

**Anything else you would like to add:**

[discussion in #cluster-api-vsphere](https://kubernetes.slack.com/archives/CKFGK3SSD/p1619621933222400)

/kind feature

### PR DESCRIPTION
**Pull request by schrej**
**What this PR does / why we need it**:
This PR adds support for annotations to the `util.CloneTemplate` and `util.GenerateTemplate` functions. This is used to pass down annotations from `MachineDeployments`/`MachineSets` to the infrastructure machine resources.

It also extends the tests of the MachineSet controller to verify that labels and annotations are passed down correctly.

**Which issue(s) this PR fixes**:
Fixes #4543

### CODE REVIEW

File: controllers/external/util_test.go [hunk 1]
```diff
@@ -175,7 +180,12 @@ func TestCloneTemplateResourceFound(t *testing.T) {
 		ClusterName: testClusterName,
 		OwnerRef:    owner.DeepCopy(),
 		Labels: map[string]string{
-			"test-label-1": "value-1",
```
Comment (enxebre) [inline controllers/external/util_test.go:178]: why are we dropping this label?

Comment (schrej) [reply controllers/external/util_test.go:178]: It is covered by the precedence label below, which is a little more clearer with regards to whether precedence is correct, but also verifies that labels are applied in general. That's why I decided to drop it, as it doesn't seem necessary to have two labels. I can re-add it if you disagree.

File: controllers/machineset_controller_test.go [hunk 2]
```diff
@@ -191,6 +199,22 @@ var _ = Describe("MachineSet Reconciler", func() {
...
+		infraMachines.SetKind("InfrastructureMachine")
+		Eventually(func() int {
+			if err := testEnv.List(ctx, infraMachines, client.InNamespace(namespace.Name)); err != nil {
+				return -1
+			}
+			return len(machines.Items)
+		}, timeout).Should(BeEquivalentTo(replicas))
+		for _, im := range infraMachines.Items {
+			Expect(im.GetAnnotations()).To(HaveKeyWithValue("annotation-1", "true"), "have annotations of MachineTemplate applied")
+			Expect(im.GetAnnotations()).To(HaveKeyWithValue("precedence", "MachineTemplate"))
```
[the remark this answers is not in the document]
Comment (schrej) [reply controllers/machineset_controller_test.go]: I've added a description

**Comment by schrej** [comment]
That test doesn't seem related, so maybe once again
/retest

**Review by vincepri — APPROVED** [review]
/retest
/lgtm
/assign @fabriziopandini @CecileRobertMichon

**Comment by schrej** [comment]
I'll have another look on friday, maybe it's related after all

**Comment by schrej** [comment]
Tests are passing locally, I also can't really see a relation. In addition the tests passed on prow before, the only thing I changed was extending tests, and those are passing.
/retest

**Review by vincepri — APPROVED** [review]
/approve
/lgtm

**Pull request merged**